File: GOF/Builder/java/src/Builder/CountingMazeBuilder.java
package Builder;

public class CountingMazeBuilder extends MazeBuilder {

	private int m_doors;
	private int m_rooms;

	public CountingMazeBuilder() {
		m_doors = 0;
		m_rooms = 0;
	}

	@Override
	public void buildMaze() {
	}

	@Override
	public void buildRoom(int n) {
		m_rooms++;
	}

	@Override
	public void buildDoor(int n1, int n2) {
		m_doors++;
	}

	public int getRoomCounts() {
		return m_rooms;
	}

	public int getDoorCounts() {
		return m_doors;
	}
}

File: GOF/Builder/cpp/builder/CountingMazeBuilder.cpp
#include "CountingMazeBuilder.h"

CountingMazeBuilder::CountingMazeBuilder()
	: m_rooms{ 0 }, m_doors{ 0 }
{}

void CountingMazeBuilder::buildMaze()
{}

void CountingMazeBuilder::buildRoom(int n)
{
	++m_rooms;
}

void CountingMazeBuilder::buildDoor(int n1, int n2)
{
	++m_doors;
}

void CountingMazeBuilder::getCounts(int &rooms, int &doors)
{
	rooms = m_rooms;
	doors = m_doors;
}